File: bluetooth test 1/update6-6-68/BluetoothManager.m
  #import "test-Bridging-Header.h"
  #import "test-Swift.h"  // Replace with your actual Swift header name

  @interface BluetoothBridge () <BluetoothScanDelegate>
  @property (nonatomic, strong) BluetoothManager *manager;
  @property (nonatomic, strong) NSDictionary *lastDevice;
  @property (nonatomic, strong) NSMutableArray *discoveredDevices;
  @property (nonatomic, strong) NSMutableArray *connectedDevices;
  @property (nonatomic, assign) BOOL isCurrentlyScanning;
  @end

  @implementation BluetoothBridge

  - (instancetype)init {
      self = [super init];
      if (self) {
          _manager = [[BluetoothManager alloc] init];
          _manager.delegate = self;
          _discoveredDevices = [[NSMutableArray alloc] init];
          _connectedDevices = [[NSMutableArray alloc] init];
          _isCurrentlyScanning = NO;
          NSLog(@"BluetoothBridge initialized with manager: %@", _manager);
      }
      return self;
  }

  #pragma mark - Basic Scanning Methods

  - (void)startBluetoothScan {
      NSLog(@"BluetoothBridge: Starting scan...");
      [_discoveredDevices removeAllObjects];
      _isCurrentlyScanning = YES;
      [_manager startScan];
  }

  - (void)stopBluetoothScan {
      NSLog(@"BluetoothBridge: Stopping scan...");
      _isCurrentlyScanning = NO;
      [_manager stopScan];
  }

  #pragma mark - Connection Methods

  - (void)connectToDeviceWithUUID:(NSString *)uuid {
      NSLog(@"BluetoothBridge: Connecting to device with UUID: %@", uuid);
      [_manager connectToPeripheralWith:uuid];
  }

  - (void)disconnectFromDeviceWithUUID:(NSString *)uuid {
      NSLog(@"BluetoothBridge: Disconnecting from device with UUID: %@", uuid);
      [_manager disconnectPeripheralWith:uuid];
  }

  - (void)reconnectToDeviceWithUUID:(NSString *)uuid {
      NSLog(@"BluetoothBridge: Reconnecting to device with UUID: %@", uuid);
      [_manager reconnectToPeripheralWith:uuid];
  }

  #pragma mark - Configuration Methods

  - (void)setAutoReconnectEnabled:(BOOL)enabled {
      NSLog(@"BluetoothBridge: Setting auto-reconnect to: %@", enabled ? @"YES" : @"NO");
      [_manager setAutoReconnect:enabled];
  }

  - (void)setMaxReconnectAttempts:(NSInteger)attempts {
      NSLog(@"BluetoothBridge: Setting max reconnect attempts to: %ld", (long)attempts);
      [_manager configureMaxReconnectAttempts:(int)attempts];  // เปลี่ยนจาก setMaxReconnectAttempts
  }

  - (void)setupAutoReconnectHandler {
      [_manager setAutoReconnectHandler:^(NSString *uuid) {
          dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(2.0 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
              [self->_manager reconnectToPeripheralWith:uuid];
          });
      }];
  }

  #pragma mark - BluetoothScanDelegate Methods

  - (void)didDiscoverDeviceWithName:(NSString *)name
                             uuid:(NSString *)uuid
                             rssi:(NSInteger)rssi
                              major:(NSInteger)major
                              minor:(NSInteger)minor {
    
      NSLog(@"BluetoothBridge: Device discovered - Name: %@, UUID: %@, RSSI: %ld, Major: %ld, Minor: %ld",
            name, uuid, (long)rssi, (long)major, (long)minor);
    
      NSDictionary *deviceInfo = @{
          @"name": name ?: @"Unknown",
          @"uuid": uuid ?: @"",
          @"rssi": @(rssi),
          @"major": @(major),
          @"minor": @(minor),
          @"timestamp": @([[NSDate date] timeIntervalSince1970])
      };
    
      self.lastDevice = deviceInfo;
    
      // Check if device already exists in discovered devices
      BOOL deviceExists = NO;
      for (NSMutableDictionary *existingDevice in _discoveredDevices) {
          if ([existingDevice[@"uuid"] isEqualToString:uuid]) {
              // Update existing device info
              [existingDevice setValuesForKeysWithDictionary:deviceInfo];
              deviceExists = YES;
              break;
          }
      }
    
      if (!deviceExists) {
          [_discoveredDevices addObject:[deviceInfo mutableCopy]];
      }
    
      NSLog(@"BluetoothBridge: Total discovered devices: %lu", (unsigned long)[_discoveredDevices count]);
  }

  - (void)didConnectToDeviceWithUUID:(NSString *)uuid {
      NSLog(@"BluetoothBridge: Successfully connected to device: %@", uuid);
    
      // Find device in discovered devices and add to connected devices
      for (NSDictionary *device in _discoveredDevices) {
          if ([device[@"uuid"] isEqualToString:uuid]) {
              NSMutableDictionary *connectedDevice = [device mutableCopy];
              connectedDevice[@"connectionStatus"] = @"connected";
              connectedDevice[@"connectedTimestamp"] = @([[NSDate date] timeIntervalSince1970]);
            
              // Remove if already exists in connected devices
              [_connectedDevices removeObject:connectedDevice];
              [_connectedDevices addObject:connectedDevice];
              break;
          }
      }
    
      NSLog(@"BluetoothBridge: Total connected devices: %lu", (unsigned long)[_connectedDevices count]);
  }

  - (void)didDisconnectFromDeviceWithUUID:(NSString *)uuid error:(NSError *)error {
      NSString *errorMessage = error ? error.localizedDescription : @"No error";
      NSLog(@"BluetoothBridge: Disconnected from device: %@ with error: %@", uuid, errorMessage);
    
      // Remove from connected devices
      NSMutableArray *devicesToRemove = [[NSMutableArray alloc] init];
      for (NSDictionary *device in _connectedDevices) {
          if ([device[@"uuid"] isEqualToString:uuid]) {
              [devicesToRemove addObject:device];
          }
      }
      [_connectedDevices removeObjectsInArray:devicesToRemove];
    
      NSLog(@"BluetoothBridge: Total connected devices: %lu", (unsigned long)[_connectedDevices count]);
  }

  - (void)didFailToConnectToDeviceWithUUID:(NSString *)uuid error:(NSError *)error {
      NSString *errorMessage = error ? error.localizedDescription : @"Unknown connection error";
      NSLog(@"BluetoothBridge: Failed to connect to device: %@ with error: %@", uuid, errorMessage);
  }

  #pragma mark - Data Retrieval Methods

  - (NSDictionary *)getLastDiscoveredDevice {
      NSLog(@"BluetoothBridge: getLastDiscoveredDevice called, returning: %@", self.lastDevice);
      return self.lastDevice;
  }

  - (NSArray *)getAllDiscoveredDevices {
      NSLog(@"BluetoothBridge: getAllDiscoveredDevices called, returning %lu devices", 
            (unsigned long)[_discoveredDevices count]);
      return [_discoveredDevices copy];
  }

  - (NSArray *)getConnectedDevices {
      NSLog(@"BluetoothBridge: getConnectedDevices called, returning %lu devices", 
            (unsigned long)[_connectedDevices count]);
      return [_connectedDevices copy];
  }

  - (NSInteger)getConnectedDevicesCount {
      NSInteger count = [_manager getConnectedPeripheralsCount];
      NSLog(@"BluetoothBridge: getConnectedDevicesCount called, returning: %ld", (long)count);
      return count;
  }

  #pragma mark - Status Methods

  - (BOOL)isScanning {
      return _isCurrentlyScanning;
  }

  - (BOOL)isBluetoothEnabled {
      // This would need to be implemented in BluetoothManager to check CBCentralManager state
      return YES; // Placeholder
  }

  - (NSString *)getBluetoothState {
      // This would need to be implemented in BluetoothManager to return actual state
      return @"poweredOn"; // Placeholder
  }

  #pragma mark - Helper Methods for Python Integration

  - (NSDictionary *)getDeviceInfoByUUID:(NSString *)uuid {
      for (NSDictionary *device in _discoveredDevices) {
          if ([device[@"uuid"] isEqualToString:uuid]) {
              return device;
          }
      }
      return nil;
  }

  - (NSString *)getDevicesAsJSONString {
      NSError *error;
      NSData *jsonData = [NSJSONSerialization dataWithJSONObject:_discoveredDevices
                                                       options:NSJSONWritingPrettyPrinted
                                                         error:&error];
      if (error) {
          NSLog(@"Error converting devices to JSON: %@", error.localizedDescription);
          return @"[]";
      }
    
      return [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];
  }

  - (NSString *)getConnectedDevicesAsJSONString {
      NSError *error;
      NSData *jsonData = [NSJSONSerialization dataWithJSONObject:_connectedDevices
                                                       options:NSJSONWritingPrettyPrinted
                                                         error:&error];
      if (error) {
          NSLog(@"Error converting connected devices to JSON: %@", error.localizedDescription);
          return @"[]";
      }
    
      return [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];
  }

  @end

File: bluetooth test 1/update6-6-68/BluetoothManager.swift
import Foundation
import CoreBluetooth

@objc protocol BluetoothScanDelegate {
    func didDiscoverDevice(name: String?, uuid: String, rssi: Int, major: Int, minor: Int)
}

@objcMembers
class BluetoothManager: NSObject, CBCentralManagerDelegate, CBPeripheralDelegate {
    var centralManager: CBCentralManager?
    
    weak var delegate: BluetoothScanDelegate?
    var discoveredPeripherals: [UUID: CBPeripheral] = [:]
    //////////
    var disconnectedPeripherals: [UUID: CBPeripheral] = [:] // เก็บ peripheral ที่ disconnect แล้ว
    private var _autoReconnect: Bool = false // เปลี่ยนเป็น private
    var maxReconnectAttempts: Int = 3
    var reconnectAttempts: [UUID: Int] = [:]
///////////
    override init() {
        super.init()
        centralManager = CBCentralManager(delegate: self, queue: nil)
    }

    func startScan() {
        guard let manager = centralManager else { return }
        if manager.state == .poweredOn {
            manager.scanForPeripherals(withServices: nil, options: nil)
            print("Swift: Scanning started")
        }
    }

    func stopScan() {
        centralManager?.stopScan()
        print("Swift: Scanning stopped")
    }

    func centralManager(_ central: CBCentralManager, didDiscover peripheral: CBPeripheral,
                        advertisementData: [String : Any], rssi RSSI: NSNumber) {
        let uuid = peripheral.identifier.uuidString
        let name = peripheral.name ?? "Unknown"
        discoveredPeripherals[peripheral.identifier] = peripheral

        var major = 0
        var minor = 0

        if let manufacturerData = advertisementData[CBAdvertisementDataManufacturerDataKey] as? Data {
            if manufacturerData.count >= 25 {
                major = Int(manufacturerData[20]) << 8 | Int(manufacturerData[21])
                minor = Int(manufacturerData[22]) << 8 | Int(manufacturerData[23])
            }
        }

        delegate?.didDiscoverDevice(name: name, uuid: uuid, rssi: RSSI.intValue, major: major, minor: minor)
    }

    func centralManagerDidUpdateState(_ central: CBCentralManager) {
        print("Bluetooth state changed: \(central.state.rawValue)")
    }
    @objc(connectToPeripheralWith:)
    func connectToPeripheral(uuidString: String) {
        guard let uuid = UUID(uuidString: uuidString),
              let peripheral = discoveredPeripherals[uuid] else {
            print("Peripheral with UUID \(uuidString) not found")
            return
        }

        centralManager?.connect(peripheral, options: nil)
    }
    func centralManager(_ central: CBCentralManager, didConnect peripheral: CBPeripheral) {
        print("✅ Connected to peripheral: \(peripheral.name ?? "Unknown")")
        //////////////////
        
        // รีเซ็ต reconnect attempts เมื่อเชื่อมต่อสำเร็จ
        reconnectAttempts[peripheral.identifier] = 0
        
        // ย้ายจาก disconnected กลับไป discovered
        if disconnectedPeripherals[peripheral.identifier] != nil {
            disconnectedPeripherals.removeValue(forKey: peripheral.identifier)
            discoveredPeripherals[peripheral.identifier] = peripheral
        }
        
        ///////////////////
        peripheral.delegate = self
        peripheral.discoverServices(nil)////////////////////
    
    }///////////////////////
//////////////////////////////////////
    func centralManager(_ central: CBCentralManager, didDisconnectPeripheral peripheral: CBPeripheral, error: Error?) {
        print("❌ Disconnected from peripheral: \(peripheral.name ?? "Unknown")")
        
        // เก็บ peripheral ที่ disconnect ไว้สำหรับ reconnect
        disconnectedPeripherals[peripheral.identifier] = peripheral
        
        // ลบ threading logic ออก - ให้ bridge จัดการ
        // Auto reconnect ถ้าเปิดใช้งาน
        if _autoReconnect {
            // แค่เรียก reconnect ตรงๆ หรือให้ bridge จัดการ timing
            let peripheralUUID = peripheral.identifier.uuidString
            // อาจจะเรียกผ่าน delegate callback แทน
            print("Auto reconnect requested for: \(peripheralUUID)")
        }
    }

    func centralManager(_ central: CBCentralManager, didFailToConnect peripheral: CBPeripheral, error: Error?) {
        print("❌ Failed to connect to peripheral: \(peripheral.name ?? "Unknown"), error: \(error?.localizedDescription ?? "Unknown")")
        
        // ลบ threading logic ออก
        if _autoReconnect {
            let peripheralUUID = peripheral.identifier.uuidString
            let currentAttempts = reconnectAttempts[peripheral.identifier] ?? 0
            if currentAttempts < maxReconnectAttempts {
                print("Auto reconnect requested for: \(peripheralUUID)")
                // ให้ bridge หรือ Python layer จัดการ timing
            }
        }
    }

    @objc(reconnectToPeripheralWith:)
    func reconnectToPeripheral(uuidString: String) {
        print("Device: \(uuidString) try to reconnect ...")
        
        guard let uuid = UUID(uuidString: uuidString) else {
            print("Invalid UUID string: \(uuidString)")
            return
        }
        
        // หา peripheral จาก discovered หรือ disconnected
        var peripheral: CBPeripheral?
        
        if let discoveredPeripheral = discoveredPeripherals[uuid] {
            peripheral = discoveredPeripheral
        } else if let disconnectedPeripheral = disconnectedPeripherals[uuid] {
            peripheral = disconnectedPeripheral
        }
        
        guard let targetPeripheral = peripheral else {
            print("Peripheral with UUID \(uuidString) not found for reconnection")
            return
        }
        
        // เช็ค reconnect attempts
        let currentAttempts = reconnectAttempts[uuid] ?? 0
        if currentAttempts >= maxReconnectAttempts {
            print("Max reconnect attempts reached for device: \(uuidString)")
            return
        }
        
        // เพิ่ม attempt count
        reconnectAttempts[uuid] = currentAttempts + 1
        
        // ทำการ reconnect
        centralManager?.connect(targetPeripheral, options: nil)
        print("Reconnection attempt \(currentAttempts + 1)/\(maxReconnectAttempts) for device: \(targetPeripheral.name ?? "Unknown")")
    }

    @objc(disconnectPeripheralWith:)
    func disconnectPeripheral(uuidString: String) {
        guard let uuid = UUID(uuidString: uuidString),
              let peripheral = discoveredPeripherals[uuid] else {
            print("Peripheral with UUID \(uuidString) not found for disconnection")
            return
        }
        
        centralManager?.cancelPeripheralConnection(peripheral)
        print("Disconnecting from peripheral: \(peripheral.name ?? "Unknown")")
    }

    @objc(enableAutoReconnect:)
    func enableAutoReconnect(_ enabled: Bool) {
        _autoReconnect = enabled
        print("Auto reconnect set to: \(enabled)")
    }

    @objc(configureMaxReconnectAttempts:)
    func configureMaxReconnectAttempts(_ attempts: Int) {
        maxReconnectAttempts = attempts
        print("Max reconnect attempts set to: \(attempts)")
    }

    @objc var isAutoReconnectEnabled: Bool {
        return _autoReconnect
    }

    @objc func getConnectedPeripheralsCount() -> Int {
        let connectedPeripherals = centralManager?.retrieveConnectedPeripherals(withServices: []) ?? []
        return connectedPeripherals.count
    }
}

@objc protocol BluetoothBridgeDelegate {
    @objc optional func didConnectToDevice(uuid: String)
    @objc optional func didDisconnectFromDevice(uuid: String, error: Error?)
    @objc optional func didFailToConnectToDevice(uuid: String, error: Error?)
}
/////////////////////////
